[Refactor] fixed openai

baberabb opened the pull request:
The `openai` model did not have the `loglikelihood` and the `loglikelihood_rolling` methods so wasn't working. Copied the absent logic from huggingface. Haven't tested it much other than to check if it's working.

They are depreciating the completion models from Jan of next year and transferring it to the `Chat Completion API` which doesn't support log probs. I'll create an issue for adding it to greedy_until.

haileyschoelkopf commented on lm_eval/models/openai_completions.py:
```diff
@@ -244,3 +275,7 @@ def _model_call(self, inps):
         # Isn't used because we override greedy_until
         raise NotImplementedError()
+
+    def loglikelihood_rolling(self, requests):
```
I believe we do need this for perplexity tasks still, like wikitext

haileyschoelkopf reviewed (approved):
Thanks so much for starting on this! A couple things still are missing to make this work fully, but if you'd like to merge this for the rest to be addressed in a subsequent PR by you/us, feel free to merge. 

I've also noticed that we assume the GPT2 tokenizer is used. We should probably instead use the relevant ones from `tiktoken.get_encoding` which have a slightly different tokenizer interface.

baberabb commented:
Hey @haileyschoelkopf. Thanks for the feedback! Added the `logliklihood_rolling` logic from `HF`. The implementation is still a bit too complicated then its need to be, especially the batching bit, but it seems all three methods are working now. Tested on wikitext, hellaswag and gsm8k. Also added tiktoken and changed the dependency to optional.

The model still doesn't support GPT3.5 and above as they all use the chat completion endpoint. Maybe I'll circle back to that later. I was initially just trying to add async, but it's not as simple as I thought. Need to read up more on that.

Added some type hints while I was going through the repo as well.

haileyschoelkopf reviewed (approved):
Thanks so much @baberabb ! Really appreciate the help, this is fantastic. 

Thank you as well for starting on the thankless task of adding typehints :)) I'd like to work toward this as well for the rest of the repo, but will probably leave it not a priority until we get a version release out.

haileyschoelkopf commented:
Oh, also, I'm surprised that GPT3.5 does not work any more--I for some reason thought that `code-davinci-002` worked with the repo, but that's very good to know. I'm probably misremembering (this might have been on the Bigscience fork), or maybe they've since changed which API those fit into.

baberabb commented:
@haileyschoelkopf you're right! I thought they were all GPT 3. Basically they are going all in on chat models from 3.5-'turbo' [onwards](https://platform.openai.com/docs/models/overview).
